File: crates/memory-testing/capture_dumps.py
from os import remove
from shutil import copy2
from sys import argv
from subprocess import Popen, run, PIPE, STDOUT
from time import sleep


def read_file_to_byte_array(file_path):
    with open(file_path, "rb") as file:
        byte_array = bytearray(file.read())
    return byte_array


def dump_process_to_bytearray(pid, output):
    run(["gcore", "-a", str(pid)], capture_output=True, check=True)
    core_file = "core." + str(pid)
    core = read_file_to_byte_array(core_file)
    copy2(core_file, output)
    remove(core_file)
    return core


if len(argv) < 3:
    print("Usage: python3 capture_dumps.py <binary_path> <output_dir>")
    exit(1)

binary_path = argv[1]
output_dir = argv[2]

print("Memory dump capture script started")

proc = Popen(binary_path, stdout=PIPE, stderr=STDOUT, stdin=PIPE, text=True)
print("Started memory testing process with PID:", proc.pid)

# Wait a bit for it to process
sleep(1)

# Dump the process before the variables are freed
initial_core = dump_process_to_bytearray(proc.pid, output_dir + "/initial_dump.bin")
print("Initial core dump file size:", len(initial_core))

proc.stdin.write(".")
proc.stdin.flush()

# Wait a bit for it to process
sleep(1)

# Dump the process after the variables are freed
final_core = dump_process_to_bytearray(proc.pid, output_dir + "/final_dump.bin")
print("Final core dump file size:", len(final_core))

# Wait for the process to finish and print the output
stdout_data, _ = proc.communicate(input=".")
print("STDOUT:", repr(stdout_data))
print("Return code:", proc.wait())

File: crates/memory-testing/src/bin/capture-dumps.rs
use std::{
    fs,
    io::{self, prelude::*},
    path::Path,
    process::{ChildStdin, ChildStdout, Command, Stdio},
};

fn dump_process_to_bytearray(pid: u32, output_dir: &Path, output_name: &Path) -> io::Result<u64> {
    let output = Command::new("gcore")
        .args(["-a", &pid.to_string()])
        .output()?;

    if !output.status.success() {
        return io::Result::Err(io::Error::new(
            io::ErrorKind::Other,
            format!("Failed to dump process: {:?}", output),
        ));
    }

    let core_path = format!("core.{}", pid);
    let output_path = output_dir.join(output_name);
    let len = fs::copy(&core_path, output_path)?;
    fs::remove_file(&core_path)?;
    Ok(len)
}

fn wait_dump_and_continue(
    stdin: &mut ChildStdin,
    stdout: &mut ChildStdout,
    id: u32,
    base_dir: &Path,
    name: &Path,
) -> Result<(), io::Error> {
    // Read the input from the process until we get the "Waiting for dump..." message
    // That way we know the process is ready to be dumped, and we don't need to just sleep a fixed
    // amount of time
    loop {
        let mut buf = [0u8; 1024];
        let read = stdout.read(&mut buf).unwrap();
        let buf_str = std::str::from_utf8(&buf[..read]).unwrap();
        if buf_str.contains("Waiting for dump...") {
            break;
        }
    }
    let dump_size = dump_process_to_bytearray(id, &base_dir.join("output"), name)?;
    println!("Got memory dump of file size: {}", dump_size);

    stdin.write_all(b".")?;
    stdin.flush()?;

    Ok(())
}

fn main() -> io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 3 {
        println!("Usage: ./capture_dumps <binary_path> <base_dir>");
        std::process::exit(1);
    }

    let binary_path = &args[1];
    let base_dir: &Path = args[2].as_ref();

    let mut proc = Command::new(binary_path)
        .arg(base_dir)
        .stdout(Stdio::piped())
        .stdin(Stdio::piped())
        .spawn()?;
    let id = proc.id();
    println!("Started memory testing process with PID: {}", id);

    let stdin = proc.stdin.as_mut().expect("Valid stdin");
    let stdout = proc.stdout.as_mut().expect("Valid stdin");

    wait_dump_and_continue(stdin, stdout, id, base_dir, "initial_dump.bin".as_ref())?;
    wait_dump_and_continue(stdin, stdout, id, base_dir, "final_dump.bin".as_ref())?;

    // Wait for the process to finish and print the output
    let output = proc.wait()?;
    println!("Return code: {}", output);

    std::process::exit(output.code().unwrap_or(1));
}
